text/0225-configuration.md

```diff
@@ -0,0 +1,411 @@
+# OpenTelemetry Configuration
+
+A new configuration interface is proposed here in the form of a configuration model, which can be expressed as a file, and validated through a published schema.
+
+## Motivation
+
+OpenTelemetry specifies code that can operate in a variety of ways based on the end-user’s desired mode of operation. This requires a configuration interface be provided to the user so they are able to communicate this information. Currently, OpenTelemetry specifies this interface in the form of the API exposed by the SDKs and environment variables. This environment variable interface is limited in the structure of information it can communicate and the primitives it can support.
+
+### Environment Variable Interface Limitations
+
+The environment variable interface suffers from the following identified limitations:
+
+* **Flat**. Structured data is only allowed by using higher-level naming or data encoding schemes. Examples of configuration limited by lack of structured configuration include:
+  * Configuring multiple span processors, periodic metric readers, or log record processors.
+  * Configuring views.
+  * Configuring arguments for parent based sampler (sampler parent is remote and sampled vs. not sampled, sampler when parent is local and sampled vs. not sampled).
+* **Runtime dependent**. Different systems expose this interface differently (Linux, BSD, Windows). This usually means unique instructions are required to properly interact with the configuration interface on different systems.
+* **Limited values**. Many systems only allow string values to be used, but OpenTelemetry specifies many configuration values other than this type. For example, OTEL_RESOURCE_ATTRIBUTES specifies a list of key value pairs to be used as resource attributes, but there is no way to specify array values, or indicate that the value should be interpreted as non-string type.
+* **Limited validation**. Validation can only be performed by the receiver, there is no meta-configuration language to validate input.
+* **Difficult to extend**. It’s difficult to anticipate the requirements of configuring custom extension components (processors, exporters, samplers, etc), and likely not practical to represent them in a flat structure. As a result, the environment variable interface is limited to components explicitly mentioned in the specification.
+
+## Explanation
+
+Using a configuration model or configuration file, users can configure all options currently available via environment variables.
+
+### Goals
```

Comment:
This doesn't say anything about static vs. dynamic nature of the configuration, is dynamic completely out of scope?

The most pertinent example of dynamic config is sampling (e.g. Jaeger's remote sampling), but technically can be extended to many other aspects, like size of internal queue in the exporter, metrics aggregation rules, etc.

Reply:
+1 for making clear if/how/when dynamic vs static configuration is in scope. Below interoperability with OpAMP is mentioned, does this mean that the application needs to be restarted or is a reconfiguration possible? 

There's probably a subset of features where dynamic is not applicable (change service_name at runtime??), but many could profit from it, beyond the ones @yurishkuro called out I can think of a few more like enabling/disabling the collection of certain attributes (e.g. python&java have OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_* and I'd like to turn them on/off if needed or some other ways of increasing "verbosity)

Reply:
Yes, dynamic is out of scope.

Reply:
My concern with this is what it means for the dynamic configs to be supported later if we do not account for the possibility from the start. It's a bit like saying "let's first design the front wheels of the car and worry about the back wheels later". We wouldn't want the dynamic configuration to use a completely different config syntax, right?

Reply:
The specification for [MeterProvider](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/metrics/sdk.md#L117-L123), [TracerProvider](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/trace/sdk.md#L70-L77), and [LoggerProvider](https://github.com/open-telemetry/opentelemetry-specification/blame/main/specification/logs/sdk.md#L55-L61) each have language allowing updating configuration.

When I think about dynamic configuration, I imagine it working as follows:
- A language would need to support reconfiguration of the providers. Maybe in the future the spec upgrades this language from a "MAY" to "SHOULD".
- Some "supervisor" code would interact with a config server. 
- The supervisor would receive config updates in the file format specified in this otep.
- The supervisor would parse the config file, and apply those changes to the providers.

The way I see it, with the addition of file based configuration, all the pieces exist for dynamic configuration. Its up to opamp specification / implementations to weave them together.

Reply:
> My concern with this is what it means for the dynamic configs to be supported later if we do not account for the possibility from the start. It's a bit like saying "let's first design the front wheels of the car and worry about the back wheels later"

Do you have any specific examples in mind where something in this proposal could conflict with a future dynamic configuration solution?

From my perspective, a fair amount of software needs to be written to enable dynamic configuration, buts it seems straight forward the components involved will be the SDKs (with support for updating config), OpAmp for coordinating updates, and file based configuration to the target configuration.

Reply:
This is related to the other unresolved discussion https://github.com/open-telemetry/oteps/pull/225#discussion_r1116286908. In the absence of a proposed API it's impossible to judge how it would affect dynamic configuration chances. For example, an API like `configObj.makeTracerProvider()` is quite difficult to make work with dynamic configuration since it assumes re-creating the provider upon configuration change, any code that already holds a reference to previous provider or tracer won't get config updates.

Reply:
I've [proposed](https://github.com/open-telemetry/oteps/pull/225/files#r1129598846) more concrete details for the `Configure` API, which states that it returns SDK components (TracerProvider, MeterProvider, LoggerProvider).

While the specification for TracerProvider, MeterProvider, and LoggerProvider allow for updating configuration, how exactly that API is structured is unspecified. I can imagine dynamic config with file config working in a couple of different ways:
- Perhaps the file configuration component provides an `update(TracerProvider, MeterProvider, LoggerProvider)` API in addition to `configure()`. The responsibility of this method would be to interpret the parsed `Configuration` model, and update the providers to reflect the configuration using whatever APIs the providers expose for updating.
- A simpler variation of this would be to explicitly define the API that the providers should expose to update. I think a good option would be for each provider to have a method called `update`, which accepts an argument called provider of the same type (i.e. `tracerProvider.update(TracerProvider)). When invoked, `update` updates the provider instance it is invoked on to reflect the configuration of its argument. If this API exists, then an OpAmp client receiving updated SDK configuration would simply call `configure(..)` to obtain providers with the new configuration, and call `update(..)` on each of the application's providers, passing the newly obtained providers as arguments. 

> For example, an API like configObj.makeTracerProvider() is quite difficult to make work with dynamic configuration since it assumes re-creating the provider upon configuration change, any code that already holds a reference to previous provider or tracer won't get config updates.

In my proposal this isn't a problem because the existing providers are updated in place.